pkg/ingester/ingester.go

```diff
@@ -2040,14 +2040,20 @@ func (i *Ingester) shipBlocks(ctx context.Context, allowed *util.AllowedTenants)
 }
 
 func (i *Ingester) compactionLoop(ctx context.Context) error {
-	ticker := time.NewTicker(i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval)
-	defer ticker.Stop()
+	// At ingester startup, spread the first compaction over the configured compaction
+	// interval. Then, the next compactions will happen at a regular interval. This logic
+	// helps to have different ingesters running the compaction at a different time,
+	// effectively spreading the compactions over the configured interval.
+	nextPeriodicCompaction := util.DurationWithNegativeJitter(i.cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval, 1)
 
 	for ctx.Err() == nil {
 		select {
-		case <-ticker.C:
+		case <-time.After(nextPeriodicCompaction):
```

Comment:
If force compaction is triggered, it will restart wait time, so subsequent compactions will be shifted (contrary to what help string says).

We could use Ticker and simply reset its duration after first compaction.

Reply:
WDYT about https://github.com/grafana/mimir/pull/4364/commits/6cb30befe97e306865358fdd84491bc48b4c095a ?